### ISSUE TITLE
"Anonymizing your MAC Address" fails for ens6/wired device.

### ISSUE THREAD

**Issue by qubesissues** [QubesOS/qubes-issues#3928]
### Qubes OS version:

4.0 Release Candidate upgraded

### Affected component(s): 

sys-net after applying https://www.qubes-os.org/doc/anonymizing-your-mac-address/

---

### Steps to reproduce the behavior:

Following the instructions in https://www.qubes-os.org/doc/anonymizing-your-mac-address/ does randomize the wireless mac address on each boot.  The ens6 device (ethernet port) never changes.

### Expected behavior:

It is expected that after applying the instructions on this page (https://www.qubes-os.org/doc/anonymizing-your-mac-address/) both my wired and wireless mac addresses will be randomized (different on each boot).

### Actual behavior:

The wireless mac changes on each boot but the wired address does not.  Seen with sudo ip link show.

### General notes:

Is this a bug with Qubes 4 or simply a documentation error?  Can anyone comfirm this problem?

---

**Comment by andrewdavidwong** [comment]
Looks like a possible duplicate of #3905.

**Comment by andrewdavidwong** [comment]
Related: #938.

**Comment by andrewdavidwong** [comment]
@qubesissues: Please see the comments on #3905, especially https://github.com/QubesOS/qubes-issues/issues/3905#issuecomment-389770543.

**Comment by qubesissues** [comment]
@andrewdavidwong It is not related to https://github.com/QubesOS/qubes-issues/issues/3905 because the responding person says "I think you are expecting the outcome to be a random MAC address per VM, which I don't think is currently possible" but that is not my goal.  I want the wireless and wired interfaces on sys-net to be randomized.  But now only the wireless is randomized based on the instructions here: https://www.qubes-os.org/doc/anonymizing-your-mac-address/ (under the heading "Randomize all Ethernet and Wifi connections")

It is not related to https://github.com/QubesOS/qubes-issues/issues/938 in any way that I see.  That issue is discussing ways to include this by default.  I am asking if the doc is wrong at https://www.qubes-os.org/doc/anonymizing-your-mac-address/ or if there is a bug in Qubes that makes this method fail.

**Comment by andrewdavidwong** [comment]
@qubesissues: Ok, I understand how your issue is distinct now. Thank you for clarifying.

**Comment by unman** [comment]
@andrewdavidwong  I've added a warning to the [Qubes page](https://www.qubes-os.org/doc/anonymizing-your-mac-address/).
This really isn't a Qubes issue though - there are numerous reports online of NM settings not being saved in the GUI and MAC anonymization not working. That would seem to be an upstream bug.

**Comment by straussmaus** [comment]
@unman @andrewdavidwong 

I was going to open a new Issue but this seems related to what I have.  I am not using the GUI instructions so the comment above (and below) does not apply to this issue:

>This really isn't a Qubes issue though - there are numerous reports online of NM settings not being saved in the GUI and MAC anonymization not working. That would seem to be an upstream bug.

I am using the config file instructions on https://www.qubes-os.org/doc/anonymizing-your-mac-address/ under section `Randomize all Ethernet and Wifi connections`

So I think it is a Qubes bug.  At a minimum the config file section also needs a warning.

**Comment by unman** [comment]
@straussmaus You dont actually say what your issue is.
Is your suggestion that when including that config snippet your MAC address is *not* changing?
I would want to see a good deal more information about your setup before accepting that claim, as well as log evidence (from the upstream router, as well as from your sys-net equivalent).

**Comment by straussmaus** [comment]
@unman

You are right.  I should have written "I have the exact problem in the first comment".

Yes, I am saying that when I include that config snippet from https://www.qubes-os.org/doc/anonymizing-your-mac-address/ as `/etc/NetworkManager/conf.d/00-macrandomize.conf` and re boot, the MAC address of `enp6` is the same after each reboot.

I tested to be consistent with the original post: `sudo ip link show` after `sudo ifconfig` and found the default/factory MAC every time.

I did not post screenshots of my router because I do not usually plug the machine into a router until after I see the MAC change with `ifconfig`

However there is an interesting new change:  I configured a new router specifically for testing this and I found an interesting thing.  When the port is connected to a DHCP server, the MAC does change.  But when I reboot with the port disconnected, it does not change.

So my hypothesis is that the MAC change must be happening at the moment a DHCP request is being made?  Or is it that the interface itself is not up until this request is made, and the config triggers the MAC change when the interface comes up?

Is that the meaning of the `[connection]` section of the config?  If that is true I think a documentation change would be helpful because it is unusual not to be able to test something before doing it, specifically when it is about security.

So *if* my hyopthesis is correct, the doc would say: "The MAC address will not change until a connection is made to an upstream server.  You may want to test this using a router or DHCP server under your control before attempting to connect to unknown networks."

**Comment by WestleyArgentum** [comment]
FWIW I can confirm I see all of the same things the original poster and @straussmaus are saying.

Using the mac randomization configuration as outlined at https://www.qubes-os.org/doc/anonymizing-your-mac-address/ I can run `sudo ip link show` and see my wireless mac address change as expected. Ethernet remains the same until I physically connect, then my wired mac address has changed.

Can anyone more familiar with the inner workings of NetworkManager confirm that the wired mac address randomization takes place before the original mac address is leaked? If so, I definitely think the "anonymizing-your-mac-address" doc should be updated to explain this behavior.

**Comment by straussmaus** [comment]
Is there any update?  This would be great information:

```
Can anyone more familiar with the inner workings of NetworkManager confirm that the wired mac address randomization takes place before the original mac address is leaked? If so, I definitely think the "anonymizing-your-mac-address" doc should be updated to explain this behavior.
```

**Comment by parulin** [comment]
This page is now part of the external docs. Is is possible to open an issue here about external docs?

**Comment by andrewdavidwong** [comment]
> This page is now part of the external docs. Is is possible to open an issue here about external docs?

I don't think it's a good idea. The Qubes OS Project is not responsible for external documentation, and such issues may not be actionable for anyone on the Qubes team. Accepting issues for external docs in the official issue tracker also blurs the line between official and unofficial documentation.
